Feature/doc reorg  managed package (#2289)
SFDO-Tooling/CumulusCI: docs/managed_package.rst

Release a Managed Package
=========================

> This section outlines how to release a first generation (1GP) Salesforce Managed Package project. Salesforce.org's Release Engineering team practices `CumulusCI Flow<TODO link>`_, which incorporates all of these steps.

Reviewer: "Managed Package project". When is it capitalized? When is it not?

Author: Good question. I think I went a little overboard. I checked current SF docs and looks like it is lower case.

Author: Let's not capitalize. As I mentioned, I think I just went overboard with the uppercase writing this up XD

Prerequisites
-------------

This section assumes:

* CumulusCI is installed on your computer.
* A Salesforce Managed Package project has been configured for use with CumulusCI.
* A packaging org is connected to CumulusCI under the name of ``packaging``.

Deploy to a Packaging Org
-------------------------

CumulusCI enables you to deploy to your ``packaging`` org.

.. warning::

    The ``ci_master`` flow runs the `uninstall_packaged_incremental<TODO>`_ task, which deploys destructive changes to the target org.

.. code-block:: console

    $ cci flow run ci_master --org packaging

.. note::

    This flow is `source format agnostic<TODO>`_.

The ``ci_master`` flow executes these tasks in the target org.

* Updates any project dependencies
* Deploys any unpackaged metadata located in the ``pre`` directory
* Deploys package metadata

> * Deploys destructive changes to remove metadata in the target org that is no longer in the local workspace

Reviewer: Should we point out the individuals task names for each of these steps?

Author: I added a tip for listing all the steps.

Create a Beta Version
---------------------

The ``release_beta`` flow groups the common tasks that must be executed for the release of a new beta version of a project.

.. code-block:: console

    $ cci flow run release_beta --org packaging

This flow typically runs against the project's ``packaging`` org where it:

* Uploads a new beta version to the ``packaging`` org
* Creates a new GitHub release tag for the new beta version

> * `Generates Release Notes<TODO#anchor>`_

Reviewer: Can't make this a natural anchor link (for the Generate Release Notes subsection) without fudging the tense on every bullet point in this section. Is there another way to manipulate the anchor link here?

Author: I think we might have to use [cross references](https://www.sphinx-doc.org/en/master/usage/restructuredtext/roles.html#cross-referencing-arbitrary-locations). I'm going to look into this more later this week or early next once we get (most of) the content out of the way.

* Syncs feature branches with the ``main`` branch

To create a new beta version for your project without the bells and whistles, use the ``upload_beta`` task:

.. code-block:: console

    $ cci task run upload_beta --name package_version 

Test a Beta Version
-------------------

The ``ci_beta`` flow installs the latest beta version of the project on a scratch org, and executes tests against it.

.. code-block:: console

    $ cci flow run ci_beta --org <TODO>

This flow is intended to be run whenever a beta release is created.

Generate Release Notes
----------------------

The ``github_release_notes`` task fetches the text from pull requests that were merged between two given tags. The task then searches for specific titles (Critical Changes, Changes, Issues Closed, New Metadata, Installation Info, and so on) in the pull request bodies, and aggregates the text together under those titles in the GitHub tag description.

To see what the release notes look like without publishing them to GitHub:

.. code-block::

>     $ cci task run github_release_notes --tag release/1.2

Reviewer: Might be prudent to add a generic placeholder for release number in command line.

Author: I'm not sure what you mean by this.

Reviewer: @Br4nd0R I'm saying that "release/1.2" is confusing because it doesn't give concrete details as an example of what you might be retrieving, and I personally have no clue what the latest release is. Neither might the reader 6 months, 1 year, 2 years from now. So we can say release/<release_number> (and then add a note below that <release_number> could be 1.1, 2.2, 3.6 and so on).

> The ``--tag`` option aggregates text from pull requests between releases 1.2 and the next most recent release.

Reviewer: I don't full understand this line. What exactly does "next most recent release" mean in the context of most recent? How far away is 1.2 from the *most* recent release? Too much room for confusion.

Reviewer 2: Yeah, this is not clear at all.

The goal is to find pull requests that were merged for inclusion in the named release tag. So if releases 1.0, 1.1, 1.2, and 1.3 exist, and the `tag` option is set to `release/1.2`, the task will find PRs that were merged between 1.1 and 1.2.

Author: ```suggestion
The ``--tag`` option indicates which release you would like to aggregate pull request text for. In the above example, this would aggregate all change notes from pull requests between the `1.2` release and the `1.1` release.
```

To see where each line in the release notes comes from, use the ``--link_pr True`` option.

.. code-block::

>     $ cci task run github_release_notes --tag release/1.2 --link_pr True

Reviewer: Is this correct?

Author: Looks good to me.

> To publish the release notes to a release tag in GitHub, use the ``--publish`` option:

Reviewer: Something I might have missed: "use the ``--publish True`` option. Then it would match the other options listed. Correct?

Reviewer 2: Yes

Release a Managed Package
=========================
This section outlines the various steps involved in releasing a first generation (1GP) Salesforce Managed Package project.
Salesforce.org's Release Engineering team practices `CumulusCI Flow<TODO link>` which incorporates all of these steps.

Prerequisites
-------------
This section assumes:

* CumulusCI is installed on your computer.
* There exists Salesforce Managed Package project configured for use with CumulusCI.

> * You have connected your packaging org to CumulusCI under the name of ``packaging``.

Author: I've update this section to include TODOs for link to pertinent sections for each of the three pre-requisites. I also tried to include what I thought would be ways to verify: a packaging org has been connected to CumulusCI and, the package's namespace is configured in CumulusCI. Let me know if you think we need more than that to verify proper setup.

I don't love the title, but included a "I don't have a Managed Package" subheading. I tried outline the steps you provided and link to the pertinent SF materials available.

Deploy to a Packaging Org
-------------------------
CumulusCI easily enables you to easily deploy to your packaging org:

.. warning::

    The ``ci_master`` flow runs the `uninstall_packaged_incremental <TODO>` task.
    This task deploys destructive changes to the target org.

.. code-block:: console

    $ cci flow run ci_master --org packaging

.. note::

    This flow is `source format agnostic <TODO>`.

The ``ci_master`` flow does the following in the target org:

#. Update any project dependencies
#. Deploy any unpackaged metadata located in the ``pre/`` directory.
#. Deploy package metadata
#. Deploy destructive changes to remove metadata in the target org that is no longer in the local workspace.

Create a Beta Version
---------------------
The ``release_beta`` flow exists to group the common tasks that we want
to execute when we release a new beta version of a project.

.. code-block:: console

    $ cci flow run release_beta --org packaging

> It is typically run against the project's packaging org, and does the following: 

Reviewer 2: "typically" sounds a bit wishy-washy here. It won't work in another org.

> * Uploads new beta version to the packaging org

Reviewer 2: ```suggestion
* Uploads new beta version from the packaging org
```

Reviewer 2: It's probably worth emphasizing that this flow assumes the package contents were already deployed using ci_master. It does not include a step to deploy them.

> * Creates a new GitHub release tag for the new beta version

Reviewer 2: maybe explain that this is needed if you want extension packages that also use CumulusCI to be able to find the latest version when this repository is listed as a dependency

> * `Generates Release Notes`_

Reviewer 2: ```suggestion
* `Generate Release Notes`_
```
(to match the heading below)

> * Syncs feature branches with the ``main/`` branch.

Reviewer 2: "syncs" is one of those words that could mean a lot of different things. It would probably be good to link to the relevant section of the CumulusCI Flow doc, and/or say _why_ this is important (so that feature branches have the latest changes from the default branch integrated)

If you just want to create a new beta version for your project,

> without the bells and whistels, you can just use the ``upload_beta`` task:

Reviewer 2: ```suggestion
without the bells and whistles, you can just use the ``upload_beta`` task:
```

.. code-block:: console

>     $ cci task run upload_beta --name package_version 

Reviewer 2: ```suggestion
    $ cci task run upload_beta --org packaging --name package_version 
```

Test a Beta Version
-------------------

> Test your latest new Beta version with:

Reviewer 2: ```suggestion
Test your latest new beta version with:
```

.. code-block:: console

    $ cci flow run ci_beta --org <TODO> 

> This flow installs the latest beta version of the project into a scratch org, and executes tests against it.

Reviewer 2: ```suggestion
This flow installs the latest beta version of the project into a scratch org, and runs Apex tests against it.
```

This flow is intended to be run every time a beta release is created.

Generate Release Notes
----------------------
The ``github_release_notes`` task fetches the text from pull requests that
were merged between two given tags. The task then searches for specific titles
(Critical Changes, Changes, Issues Closed, New Metadata, Installation Info) in
the pull request bodies and aggregates the text together under those titles in

> in the description of a GitHub tag.

Reviewer 2: It'd probably be helpful to give an example. Is there anything in the old docset that we can use?

.. code-block::

    $ cci task run github_release_notes --tag release/1.2

This would aggregate text from pull requests between releases 1.2 and next most recent release.
You can also see where each line in the release notes comes from from by using the ``--link_pr True`` option.

If you want to publish these release notes to a release tag in GitHub use the ``--publish`` option:

.. code-block::

    $ cci task run github_release_notes --tag release/1.2 --publish True

If your team wants to use additional headings you can add new ones by
putting the following under the ``project`` -> ``git`` section of your ``cumulusci.yml`` file:

.. code-block::

    release_notes:
        parsers:
            7: class_path: cumulusci.tasks.release_notes.parser.GithubLinesParser

.. note::

    The new parser is listed with the number ``7`` because the first six are the
    `default parsers <https://github.com/SFDO-Tooling/CumulusCI/blob/671a0e88cef79e9aeefe1e2b835816cd8141bdbb/cumulusci/cumulusci.yml#L1154>`_ that come with CumulusCI.
        

Upload and Test a Final Version
-------------------------------

> When you're ready to upload a production release of your Managed Package project use the ``--production True`` option:

Reviewer 2: This is confusing because this option is not actually part of the command shown here (it's configured as part of the release_production flow)

.. code-block::

    $ cci flow run release_production --org packaging 

Similar to ``release_beta``, this task uploads a new production version of your package,

> creates a release tag in GitHub, and aggregates release notes for the new version.

Reviewer 2: Again, emphasize that it assumes the package contents were already deployed using ci_master?

If you would just like to upload the new production version 
without creating the GitHub tag and generating release notes, use:

.. code-block::

    $ cci task run upload_beta --name v1.2.1 --production True

You can test the new package version with:

.. code-block::

    $ cci flow run ci_release

> This flow installs the latest production release version and runs the tests from the managed package in a scratch org.

Reviewer 2: ```suggestion
This flow installs the latest production release version and runs the Apex tests from the managed package in a scratch org.
```
